_posts/2017-07-04-flink-rescalable-state.md

```diff
@@ -0,0 +1,172 @@
+
+As a generalized approach to solve this black box problem, we slightly modified the checkpointing interface, called `ListCheckpointed`. **Figure 2B** shows the new checkpointing interface, which returns and receives a list of state partitions. Introducing a list instead of a single object makes the meaningful partitioning of state explicit: each item in the list still remains a black box to Flink, but is considered an atomic, independently re-distributable part of the operator state.
+
+
+<br><center><i>Figure 2</i></center>
+<center>
+<img src="{{ site.baseurl }}/img/blog/list-checkpointed.svg" style="width:70%;margin:10px">
+</center><br>
+
+
+Our approach provides a simple API with which implementing operators can encode domain-specific knowledge about how to partition and merge units of state. With our new checkpointing interface, the Kafka source makes individual partition offsets explicit, and state reassignment becomes as easy as splitting and merging lists.
+
+```java
+public class FlinkKafkaConsumer<T> extends RichParallelSourceFunction<T> implements CheckpointedFunction {
+	 // ...
+
+   private transient ListState<Tuple2<KafkaTopicPartition, Long>> offsetsOperatorState;
+
+   @Override
+   public void initializeState(FunctionInitializationContext context) throws Exception {
+
+      OperatorStateStore stateStore = context.getOperatorStateStore();
+      // register the state with the backend
+      this.offsetsOperatorState = stateStore.getSerializableListState("kafka-offsets");
+
+      // if the job was restarted, we set the restored offsets
+      if (context.isRestored()) {
+         for (Tuple2<KafkaTopicPartition, Long> kafkaOffset : offsetsOperatorState.get()) {
+            // ... restore logic
+         }
+      }
+   }
+
+   @Override
+   public void snapshotState(FunctionSnapshotContext context) throws Exception {
+
+      this.offsetsOperatorState.clear();
+
+      // write the partition offsets to the list of operator states
+      for (Map.Entry<KafkaTopicPartition, Long> partition : this.subscribedPartitionOffsets.entrySet()) {
+         this.offsetsOperatorState.add(Tuple2.of(partition.getKey(), partition.getValue()));
+      }
+   }
+
+   // ...
+
+}
+```
+
+## Reassigning Keyed State When Rescaling
+The second flavour of state in Flink is keyed state. In contrast to operator state, keyed state is scoped by key, where the key is extracted from each stream event.
+
+To illustrate how keyed state differs from operator state, let’s use the following example. Assume we have a stream of events, where each event has the schema `{customer_id:int, value:int}`. We have already learned that we can use operator state to compute and emit the running sum of values for all customers.
+
+Now assume we want to slightly modify our goal and compute a running sum of values for each individual customer_id. This is a use case from keyed state, as one aggregated state must be maintained for each unique key in the stream.
```

Comment:
make "customer_id" be formatted as code, i.e. surround it with `'s => `customer_id`.